File: +fs/+lpew/+v2/lambdaWeights.m
function [lambda, r] = lambdaWeights(FS, Kt1, Kt2, Kn1, Kn2, theta1, theta2, ve1, ve2, netas, T_all, Q_corner)
%FS.LPEW.V2.LAMBDAWEIGHTS  Batched LPEW2 λ + r (replaces Lamdas_Weights_LPEW2).
%
%   [lambda, r] = fs.lpew.v2.lambdaWeights(FS, Kt1, Kt2, Kn1, Kn2, ...
%                                          theta1, theta2, ve1, ve2, netas, T, Qcorner)
%
%   Inputs (all per-corner from earlier PR-B2/B3 kernels):
%     Kt1  [nCorners x 2]     Kt2  [nCorners x 1]
%     Kn1  [nCorners x 2]     Kn2  [nCorners x 1]
%     theta1,2,ve1,2          [nCorners x 1]
%     netas                   [nCorners x 2]
%     T_all                   [nAllNeighbors x 3]  (needed for boundary r)
%     Q_corner                [nCorners x 3]       (needed for boundary r)
%
%   Outputs:
%     lambda  [nCorners x 1]  per-corner weight (interior formula fully vectorized;
%                             boundary handled per-node via legacy path)
%     r       [nNodes   x 2]  boundary Neumann correction (0 for interior nodes)
%
%   PR-B4 scope: interior nodes vectorized (majority — 80% of typical meshes).
%   Boundary nodes fall back to a per-node loop that mirrors the legacy formula.
%   Full boundary vectorization is PR-B4b (may be folded into PR-B5).

    nCorners = FS.csr.nCorners;
    nNodes   = FS.mesh.nNodes;

    lambda = zeros(nCorners, 1);
    r      = zeros(nNodes, 2);

    % ────────────────────────────────────────────────────────────────
    % INTERIOR NODES — fully vectorized (per-corner arithmetic)
    % ────────────────────────────────────────────────────────────────
    % For each corner i (in an interior node), let:
    %   iprev = FS.csr.cornerPrev(i)   (k-1 with wrap to nec)
    %   inext = FS.csr.cornerNext(i)   (k+1 with wrap to 1)
    %
    %   zeta_num(i) = Kn2(iprev)*cot(ve1(iprev)) + Kn2(i)*cot(ve2(i)) + Kt2(iprev) - Kt2(i)
    %   zeta_den(i) = Kn1(iprev,2)*cot(theta2(iprev)) + Kn1(i,1)*cot(theta1(i))
    %                 - Kt1(iprev,2) + Kt1(i,1)
    %   zeta(i)     = zeta_num / zeta_den
    %
    %   lambda(i)   = Kn1(i,1)*netas(i,1)*zeta(i) + Kn1(i,2)*netas(i,2)*zeta(inext)

    isInterior = FS.csr.nodeIsInterior(FS.csr.cornerNode);   % [nCorners x 1]
    iprev = FS.csr.cornerPrev;
    inext = FS.csr.cornerNext;

    % Compute zeta for ALL corners (interior formula) — we overwrite boundary later
    cot_ve1_prev    = cot(ve1(iprev));
    cot_ve2_cur     = cot(ve2);
    cot_theta2_prev = cot(theta2(iprev));
    cot_theta1_cur  = cot(theta1);

    zeta_num = Kn2(iprev) .* cot_ve1_prev + Kn2 .* cot_ve2_cur + Kt2(iprev) - Kt2;
    zeta_den = Kn1(iprev, 2) .* cot_theta2_prev + Kn1(:, 1) .* cot_theta1_cur ...
               - Kt1(iprev, 2) + Kt1(:, 1);
    zeta = zeta_num ./ zeta_den;

    % lambda for interior corners
    lambda_int = Kn1(:, 1) .* netas(:, 1) .* zeta ...
               + Kn1(:, 2) .* netas(:, 2) .* zeta(inext);
    lambda(isInterior) = lambda_int(isInterior);

    % ────────────────────────────────────────────────────────────────
    % BOUNDARY NODES — legacy per-node fallback (PR-B4b will vectorize)
    % ────────────────────────────────────────────────────────────────
    boundaryNodes = find(~FS.csr.nodeIsInterior);
    esurn2  = FS.mesh.esurn2;
    nsurn2  = FS.mesh.nsurn2;
    coord   = FS.mesh.coord;

    for k = 1:numel(boundaryNodes)
        No  = boundaryNodes(k);
        e_range = esurn2(No)+1 : esurn2(No+1);
        p_range = nsurn2(No)+1 : nsurn2(No+1);
        nec = numel(e_range);
        Qo  = coord(No, :);

        % Slice per-node inputs from the batched arrays
        Kt1_n = Kt1(e_range, :);
        Kt2_n = Kt2(e_range);
        Kn1_n = Kn1(e_range, :);
        Kn2_n = Kn2(e_range);
        th1_n = theta1(e_range).';
        th2_n = theta2(e_range).';
        ve1_n = ve1(e_range).';
        ve2_n = ve2(e_range).';
        netas_n = netas(e_range, :);
        T_n = T_all(p_range, :);

        % Legacy boundary formula for zeta (k=1..nec+1)
        zeta_n = zeros(1, nec+1);
        for kk = 1:nec+1
            if kk == 1
                zn = Kn2_n(kk)*cot(ve2_n(kk)) - Kt2_n(kk);
                zd = Kn1_n(kk, 1)*cot(th1_n(kk)) + Kt1_n(kk, 1);
                r(No, 1) = (1 + zn/zd) * norm(Qo - T_n(1, :));
            elseif kk == nec+1
                zn = Kn2_n(kk-1)*cot(ve1_n(kk-1)) + Kt2_n(kk-1);
                zd = Kn1_n(kk-1, 2)*cot(th2_n(kk-1)) - Kt1_n(kk-1, 2);
                r(No, 2) = (1 + zn/zd) * norm(Qo - T_n(nec+1, :));
            else
                zn = Kn2_n(kk-1)*cot(ve1_n(kk-1)) + Kn2_n(kk)*cot(ve2_n(kk)) ...
                     + Kt2_n(kk-1) - Kt2_n(kk);
                zd = Kn1_n(kk-1, 2)*cot(th2_n(kk-1)) + Kn1_n(kk, 1)*cot(th1_n(kk)) ...
                     - Kt1_n(kk-1, 2) + Kt1_n(kk, 1);
            end
            zeta_n(kk) = zn / zd;
        end

        % Lambda for boundary corners
        for kk = 1:nec
            % Boundary: always uses zeta(kk) and zeta(kk+1) (no wrap)
            lambda(e_range(kk)) = Kn1_n(kk, 1)*netas_n(kk, 1)*zeta_n(kk) ...
                                 + Kn1_n(kk, 2)*netas_n(kk, 2)*zeta_n(kk+1);
        end
    end
end

File: fs/lpew/v2/lambdaWeights.py
"""LPEW2 lambda weights translated from ``+fs/+lpew/+v2/lambdaWeights.m``."""

from typing import Any, Dict, Tuple

import numpy as np


def lambdaWeights(
    FS: Dict[str, Any],
    Kt1: np.ndarray,
    Kt2: np.ndarray,
    Kn1: np.ndarray,
    Kn2: np.ndarray,
    theta1: np.ndarray,
    theta2: np.ndarray,
    ve1: np.ndarray,
    ve2: np.ndarray,
    netas: np.ndarray,
    T_all: np.ndarray,
    Q_corner: np.ndarray,
) -> Tuple[np.ndarray, np.ndarray]:
    """Compute per-corner lambda values and boundary Neumann corrections."""
    mesh = FS["mesh"]
    csr = FS["csr"]
    n_corners = int(csr.get("nCorners", len(csr["cornerNode"])))
    n_nodes = int(mesh["nNodes"])

    Kt1 = np.asarray(Kt1, dtype=float)
    Kt2 = np.asarray(Kt2, dtype=float).reshape(-1)
    Kn1 = np.asarray(Kn1, dtype=float)
    Kn2 = np.asarray(Kn2, dtype=float).reshape(-1)
    theta1 = np.asarray(theta1, dtype=float).reshape(-1)
    theta2 = np.asarray(theta2, dtype=float).reshape(-1)
    ve1 = np.asarray(ve1, dtype=float).reshape(-1)
    ve2 = np.asarray(ve2, dtype=float).reshape(-1)
    netas = np.asarray(netas, dtype=float)
    T_all = np.asarray(T_all, dtype=float)
    Q_corner = np.asarray(Q_corner, dtype=float)

    corner_node = np.asarray(csr["cornerNode"], dtype=int)
    corner_prev = np.asarray(csr["cornerPrev"], dtype=int)
    corner_next = np.asarray(csr["cornerNext"], dtype=int)
    node_is_interior = np.asarray(csr["nodeIsInterior"], dtype=bool)
    interior_corners = node_is_interior[corner_node]

    lambda_values = np.zeros(n_corners, dtype=float)
    r = np.zeros((n_nodes, 2), dtype=float)

    zeta_num = (
        Kn2[corner_prev] * _cot(ve1[corner_prev])
        + Kn2 * _cot(ve2)
        + Kt2[corner_prev]
        - Kt2
    )
    zeta_den = (
        Kn1[corner_prev, 1] * _cot(theta2[corner_prev])
        + Kn1[:, 0] * _cot(theta1)
        - Kt1[corner_prev, 1]
        + Kt1[:, 0]
    )
    zeta = zeta_num / zeta_den
    lambda_interior = (
        Kn1[:, 0] * netas[:, 0] * zeta
        + Kn1[:, 1] * netas[:, 1] * zeta[corner_next]
    )
    lambda_values[interior_corners] = lambda_interior[interior_corners]

    esurn2 = np.asarray(mesh["esurn2"], dtype=int)
    nsurn2 = np.asarray(mesh["nsurn2"], dtype=int)
    coord = np.asarray(mesh["coord"], dtype=float)

    for node in np.flatnonzero(~node_is_interior):
        corner_slice = slice(esurn2[node], esurn2[node + 1])
        neighbor_slice = slice(nsurn2[node], nsurn2[node + 1])
        corner_indices = np.arange(corner_slice.start, corner_slice.stop)
        n_node_corners = corner_indices.size

        kt1_node = Kt1[corner_slice]
        kt2_node = Kt2[corner_slice]
        kn1_node = Kn1[corner_slice]
        kn2_node = Kn2[corner_slice]
        theta1_node = theta1[corner_slice]
        theta2_node = theta2[corner_slice]
        ve1_node = ve1[corner_slice]
        ve2_node = ve2[corner_slice]
        netas_node = netas[corner_slice]
        t_node = T_all[neighbor_slice]

        if t_node.shape[0] != n_node_corners + 1:
            raise ValueError(
                f"Boundary node {node} must have one more neighbor than corner"
            )

        zeta_node = np.empty(n_node_corners + 1, dtype=float)
        first_num = kn2_node[0] * _cot(ve2_node[0]) - kt2_node[0]
        first_den = kn1_node[0, 0] * _cot(theta1_node[0]) + kt1_node[0, 0]
        zeta_node[0] = first_num / first_den
        r[node, 0] = (1.0 + zeta_node[0]) * np.linalg.norm(
            coord[node] - t_node[0]
        )

        for local_corner in range(1, n_node_corners):
            previous = local_corner - 1
            numerator = (
                kn2_node[previous] * _cot(ve1_node[previous])
                + kn2_node[local_corner] * _cot(ve2_node[local_corner])
                + kt2_node[previous]
                - kt2_node[local_corner]
            )
            denominator = (
                kn1_node[previous, 1] * _cot(theta2_node[previous])
                + kn1_node[local_corner, 0] * _cot(theta1_node[local_corner])
                - kt1_node[previous, 1]
                + kt1_node[local_corner, 0]
            )
            zeta_node[local_corner] = numerator / denominator

        last_num = kn2_node[-1] * _cot(ve1_node[-1]) + kt2_node[-1]
        last_den = kn1_node[-1, 1] * _cot(theta2_node[-1]) - kt1_node[-1, 1]
        zeta_node[-1] = last_num / last_den
        r[node, 1] = (1.0 + zeta_node[-1]) * np.linalg.norm(
            coord[node] - t_node[-1]
        )

        next_zeta = np.concatenate((zeta_node[1:n_node_corners], zeta_node[:1]))
        lambda_values[corner_slice] = (
            kn1_node[:, 0] * netas_node[:, 0] * zeta_node[:n_node_corners]
            + kn1_node[:, 1] * netas_node[:, 1] * next_zeta
        )

    return lambda_values, r


def _cot(values: np.ndarray) -> np.ndarray:
    return 1.0 / np.tan(values)
